src/fpm_targets.f90

```diff
@@ -248,38 +272,67 @@ subroutine build_target_list(targets,model)
 
     contains
 
-    function get_object_name(source) result(object_file)
+    subroutine get_object_name(source, object_file, module_flags)
```

Comment:
I think `get_object_name` should be used for all target types to centralise the logic. At the moment, executables and archives have separate logic interspersed in the main code which makes it more difficult to reason about. I would suggest modify `get_object_name` to detect the source type and use this to build the object name if it differs between source types.